README.md
kivy-garden Matplotlib flower
========================

[![Github Build Status](https://github.com/kivy-garden/matplotlib/workflows/Garden%20flower/badge.svg)](https://github.com/kivy-garden/matplotlib/actions)

A kivy garden flower providing a matplotlib widget.

See https://kivy-garden.github.io/matplotlib/ for the rendered flower docs.

Please see the garden [instructions](https://kivy-garden.github.io) for how to use kivy garden flowers.

CI
--

Every push or pull request run the [GitHub Action](https://github.com/kivy-garden/matplotlib/actions) CI.
It tests the code on various OS and also generates wheels that can be released on PyPI upon a
tag. Docs are also generated and uploaded to the repo as well as artifacts of the CI.


Contributing
--------------

Check out our [contribution guide](CONTRIBUTING.md) and feel free to improve the flower.

License
---------

This software is released under the terms of the MIT License.
Please see the [LICENSE.txt](LICENSE.txt) file.

How to release
===============

See the garden [instructions](https://kivy-garden.github.io/#makingareleaseforyourflower) for how to make a new release.


CHANGELOG.md
# Changelog

## [Unreleased] - yyyy-mm-dd
 - Migrated to the new kivy-garden flower structure.


kivy_garden/matplotlib/tests/test_backend_kivy.py
def test_matplotlib_use_backend():
    import matplotlib
    import matplotlib.pyplot as plt

    matplotlib.use("module://kivy_garden.matplotlib.backend_kivy")

    plt.plot([1, 2, 3, 4])
    plt.ylabel("some numbers")

    plt.show(block=False)


examples/plt.py
#!/usr/bin/env python
# a bar plot with errorbars
import matplotlib

matplotlib.use("module://kivy_garden.matplotlib.backend_kivy")
# matplotlib.use('Gtk')

import numpy as np
import matplotlib.pyplot as plt


def press(event):
    print("press released from test", event.x, event.y, event.button)


def release(event):
    print("release released from test", event.x, event.y, event.button)


def keypress(event):
    print("key down", event.key)


def keyup(event):
    print("key up", event.key)


def motionnotify(event):
    print("mouse move to ", event.x, event.y)


def resize(event):
    print("resize from mpl ", event)


def scroll(event):
    print("scroll event from mpl ", event.x, event.y, event.step)


def figure_enter(event):
    print("figure enter mpl")


def figure_leave(event):
    print("figure leaving mpl")


def close(event):
    print("closing figure")


N = 5
menMeans = (20, 35, 30, 35, 27)
menStd = (2, 3, 4, 1, 2)

ind = np.arange(N)  # the x locations for the groups
width = 0.35  # the width of the bars

figure, ax = plt.subplots()

figure.canvas.mpl_connect("button_press_event", press)
figure.canvas.mpl_connect("button_release_event", release)
figure.canvas.mpl_connect("key_press_event", keypress)
figure.canvas.mpl_connect("key_release_event", keyup)
figure.canvas.mpl_connect("motion_notify_event", motionnotify)
figure.canvas.mpl_connect("resize_event", resize)
figure.canvas.mpl_connect("scroll_event", scroll)
figure.canvas.mpl_connect("figure_enter_event", figure_enter)
figure.canvas.mpl_connect("figure_leave_event", figure_leave)
figure.canvas.mpl_connect("close_event", close)

fig1 = plt.gcf()
rects1 = ax.bar(ind, menMeans, width, color="r", yerr=menStd)

womenMeans = (25, 32, 34, 20, 25)
womenStd = (3, 5, 2, 3, 3)
rects2 = ax.bar(ind + width, womenMeans, width, color="y", yerr=womenStd)

# add some text for labels, title and axes ticks
ax.set_ylabel("---------------------------Scores----------------------------")
ax.set_title("Scores by group and gender")
ax.set_xticks(ind + width)
ax.set_yticklabels(
    ("Ahh", "--G1--", "G2", "G3", "G4", "G5", "G5", "G5", "G5"), rotation=90
)
ax.legend((rects1[0], rects2[0]), ("Men", "Women"))


def autolabel(rects):
    # attach some text labels
    for rect in rects:
        height = rect.get_height()
        ax.text(
            rect.get_x() + rect.get_width() / 2.0,
            1.05 * height,
            "%d" % int(height),
            ha="center",
            va="bottom",
        )


autolabel(rects1)
autolabel(rects2)


plt.draw()
# fig1.set_size_inches(18.5, 10.5, forward = True)
plt.savefig("test.png")
plt.show()


examples/events.py
from kivy.app import App
from kivy.lang import Builder
from kivy.uix.boxlayout import BoxLayout

import numpy as np
import matplotlib.pyplot as plt
from kivy_garden.matplotlib.backend_kivyagg import FigureCanvas


def enter_axes(event):
    print("enter_axes", event.inaxes)
    event.inaxes.patch.set_facecolor("yellow")
    event.canvas.draw()


def leave_axes(event):
    print("leave_axes", event.inaxes)
    event.inaxes.patch.set_facecolor("white")
    event.canvas.draw()


def enter_figure(event):
    print("enter_figure", event.canvas.figure)
    event.canvas.figure.patch.set_facecolor("red")
    event.canvas.draw()


def leave_figure(event):
    print("leave_figure", event.canvas.figure)
    event.canvas.figure.patch.set_facecolor("grey")
    event.canvas.draw()


kv = """
<Test>:
    orientation: 'vertical'
    Button:
        size_hint_y: None
        height: 40
"""

Builder.load_string(kv)


class Test(BoxLayout):
    def __init__(self, *args, **kwargs):
        super(Test, self).__init__(*args, **kwargs)
        self.add_plot()

    def get_fc(self, i):
        fig1 = plt.figure()
        fig1.suptitle(
            "mouse hover over figure or axes to trigger events" + str(i)
        )
        ax1 = fig1.add_subplot(211)
        ax2 = fig1.add_subplot(212)
        wid = FigureCanvas(fig1)
        fig1.canvas.mpl_connect("figure_enter_event", enter_figure)
        fig1.canvas.mpl_connect("figure_leave_event", leave_figure)
        fig1.canvas.mpl_connect("axes_enter_event", enter_axes)
        fig1.canvas.mpl_connect("axes_leave_event", leave_axes)
        return wid

    def add_plot(self):
        self.add_widget(self.get_fc(1))
        self.add_widget(self.get_fc(2))


class TestApp(App):
    def build(self):
        return Test()


if __name__ == "__main__":
    TestApp().run()


kivy_garden/matplotlib/backend_kivyagg.py
"""
Backend KivyAgg
=====

.. image:: images/backend_agg_example.jpg
    :align: right

The :class:`FigureCanvasKivyAgg` widget is used to create a matplotlib graph.
The render will cover the whole are of the widget unless something different is
specified using a :meth:`blit`.
When you are creating a FigureCanvasKivyAgg widget, you must at least
initialize it with a matplotlib figure object. This class uses agg to get a
static image of the plot and then the image is render using a
:class:`~kivy.graphics.texture.Texture`. See backend_kivy documentation for
more information since both backends can be used in the exact same way.


Examples
--------

Example of a simple Hello world matplotlib App::

    fig, ax = plt.subplots()
    ax.text(0.6, 0.5, "hello", size=50, rotation=30.,
            ha="center", va="center",
            bbox=dict(boxstyle="round",
                      ec=(1., 0.5, 0.5),
                      fc=(1., 0.8, 0.8),
                      )
            )
    ax.text(0.5, 0.4, "world", size=50, rotation=-30.,
            ha="right", va="top",
            bbox=dict(boxstyle="square",
                      ec=(1., 0.5, 0.5),
                      fc=(1., 0.8, 0.8),
                      )
            )
    canvas = FigureCanvasKivyAgg(figure=fig)

The object canvas can be added as a widget into the kivy tree widget.
If a change is done on the figure an update can be performed using
:meth:`~kivy.ext.mpl.backend_kivyagg.FigureCanvasKivyAgg.draw`.::

    # update graph
    canvas.draw()

The plot can be exported to png with
:meth:`~kivy.ext.mpl.backend_kivyagg.FigureCanvasKivyAgg.print_png`, as an
argument receives the `filename`.::

    # export to png
    canvas.print_png("my_plot.png")


Backend KivyAgg Events
-----------------------

The events available are the same events available from Backend Kivy.::

    def my_callback(event):
        print('press released from test', event.x, event.y, event.button)

    fig.canvas.mpl_connect('mpl_event', my_callback)

"""

from __future__ import (
    absolute_import,
    division,
    print_function,
    unicode_literals,
)

__all__ = "FigureCanvasKivyAgg"

import six

import matplotlib
from matplotlib._pylab_helpers import Gcf
from matplotlib.backend_bases import (
    RendererBase,
    GraphicsContextBase,
    FigureManagerBase,
    FigureCanvasBase,
)
from matplotlib.figure import Figure
from matplotlib.transforms import Bbox
from matplotlib.backends.backend_agg import FigureCanvasAgg
from matplotlib.backend_bases import register_backend, ShowBase

try:
    import kivy
except ImportError:
    raise ImportError("this backend requires Kivy to be installed.")

from kivy.app import App
from kivy.graphics.texture import Texture
from kivy.graphics import Rectangle, Color
from kivy.uix.widget import Widget
from kivy.properties import ObjectProperty
from kivy.base import EventLoop
from kivy.uix.floatlayout import FloatLayout
from kivy.core.image import Image
from kivy_garden.matplotlib.backend_kivy import (
    FigureCanvasKivy,
    FigureManagerKivy,
    NavigationToolbar2Kivy,
)

register_backend("png", "backend_kivyagg", "PNG File Format")

toolbar = None
my_canvas = None


def new_figure_manager(num, *args, **kwargs):
    """Create a new figure manager instance for the figure given."""
    # if a main-level app must be created, this (and
    # new_figure_manager_given_figure) is the usual place to
    # do it -- see backend_wx, backend_wxagg and backend_tkagg for
    # examples. Not all GUIs require explicit instantiation of a
    # main-level app (egg backend_gtk, backend_gtkagg) for pylab
    FigureClass = kwargs.pop("FigureClass", Figure)
    thisFig = FigureClass(*args, **kwargs)
    return new_figure_manager_given_figure(num, thisFig)


def new_figure_manager_given_figure(num, figure):
    """Create a new figure manager instance and a new figure canvas instance
    for the given figure.
    """
    canvas = FigureCanvasKivyAgg(figure)
    manager = FigureManagerKivy(canvas, num)
    global my_canvas
    global toolbar
    toolbar = manager.toolbar.actionbar if manager.toolbar else None
    my_canvas = canvas
    return manager


class MPLKivyApp(App):
    """Creates the App initializing a FloatLayout with a figure and toolbar
    widget.
    """

    figure = ObjectProperty(None)
    toolbar = ObjectProperty(None)

    def build(self):
        EventLoop.ensure_window()
        layout = FloatLayout()
        if self.figure:
            self.figure.size_hint_y = 0.9
            layout.add_widget(self.figure)
        if self.toolbar:
            self.toolbar.size_hint_y = 0.1
            layout.add_widget(self.toolbar)
        return layout


class Show(ShowBase):
    """mainloop needs to be overwritten to define the show() behavior for kivy
    framework.
    """

    @classmethod
    def mainloop(self):
        global my_canvas
        global toolbar
        app = App.get_running_app()
        if app is None:
            app = MPLKivyApp(figure=my_canvas, toolbar=toolbar)
            app.run()


show = Show()


class FigureCanvasKivyAgg(FigureCanvasKivy, FigureCanvasAgg):
    """FigureCanvasKivyAgg class. See module documentation for more
    information.
    """

    def __init__(self, figure, **kwargs):
        self.figure = figure
        self.bind(size=self._on_size_changed)
        super(FigureCanvasKivyAgg, self).__init__(figure=self.figure, **kwargs)
        self.img_texture = None
        self.img_rect = None
        self.blit()

    def draw(self):
        """
        Draw the figure using the agg renderer
        """
        self.canvas.clear()
        FigureCanvasAgg.draw(self)
        if self.blitbox is None:
            l, b, w, h = self.figure.bbox.bounds
            w, h = int(w), int(h)
            buf_rgba = self.get_renderer().buffer_rgba()
        else:
            bbox = self.blitbox
            l, b, r, t = bbox.extents
            w = int(r) - int(l)
            h = int(t) - int(b)
            t = int(b) + h
            reg = self.copy_from_bbox(bbox)
            buf_rgba = reg.to_string()
        texture = Texture.create(size=(w, h))
        texture.flip_vertical()
        color = self.figure.get_facecolor()
        with self.canvas:
            Color(*color)
            Rectangle(pos=self.pos, size=(w, h))
            Color(1.0, 1.0, 1.0, 1.0)
            self.img_rect = Rectangle(
                texture=texture, pos=self.pos, size=(w, h)
            )
        texture.blit_buffer(bytes(buf_rgba), colorfmt="rgba", bufferfmt="ubyte")
        self.img_texture = texture

    filetypes = FigureCanvasKivy.filetypes.copy()
    filetypes["png"] = "Portable Network Graphics"

    def _on_pos_changed(self, *args):
        if self.img_rect is not None:
            self.img_rect.pos = self.pos

    def _print_image(self, filename, *args, **kwargs):
        """Write out format png. The image is saved with the filename given."""
        l, b, w, h = self.figure.bbox.bounds
        img = None
        if self.img_texture is None:
            texture = Texture.create(size=(w, h))
            texture.blit_buffer(
                bytes(self.get_renderer().buffer_rgba()),
                colorfmt="rgba",
                bufferfmt="ubyte",
            )
            texture.flip_vertical()
            img = Image(texture)
        else:
            img = Image(self.img_texture)
        img.save(filename)


""" Standard names that backend.__init__ is expecting """
FigureCanvas = FigureCanvasKivyAgg
FigureManager = FigureManagerKivy
NavigationToolbar = NavigationToolbar2Kivy
show = show


examples/backend.py
import matplotlib

matplotlib.use("module://kivy_garden.matplotlib.backend_kivy")
from matplotlib.figure import Figure
from numpy import arange, sin, pi
from kivy.app import App

import numpy as np
from kivy_garden.matplotlib.backend_kivyagg import (
    FigureCanvas,
    NavigationToolbar2Kivy,
)
from kivy.uix.floatlayout import FloatLayout
from kivy.uix.boxlayout import BoxLayout
from matplotlib.transforms import Bbox
from kivy.uix.button import Button
from kivy.graphics import Color, Line, Rectangle

import matplotlib.pyplot as plt


def press(event):
    print("press released from test", event.x, event.y, event.button)


def release(event):
    print("release released from test", event.x, event.y, event.button)


def keypress(event):
    print("key down", event.key)


def keyup(event):
    print("key up", event.key)


def motionnotify(event):
    print("mouse move to ", event.x, event.y)


def resize(event):
    print("resize from mpl ", event.width, event.height)


def scroll(event):
    print("scroll event from mpl ", event.x, event.y, event.step)


def figure_enter(event):
    print("figure enter mpl")


def figure_leave(event):
    print("figure leaving mpl")


def close(event):
    print("closing figure")


N = 5
menMeans = (20, 35, 30, 35, 27)
menStd = (2, 3, 4, 1, 2)

ind = np.arange(N)  # the x locations for the groups
width = 0.35  # the width of the bars

fig, ax = plt.subplots()
fig2, ax2 = plt.subplots()
rects1 = ax.bar(ind, menMeans, width, color="r", yerr=menStd)

womenMeans = (25, 32, 34, 20, 25)
womenStd = (3, 5, 2, 3, 3)
rects2 = ax.bar(ind + width, womenMeans, width, color="y", yerr=womenStd)

# add some text for labels, title and axes ticks
ax.set_ylabel("Scores")
ax.set_title("Scores by group and gender")
ax.set_xticks(ind + width)
ax.set_xticklabels(("G1", "G2", "G3", "G4", "G5"))
ax.legend((rects1[0], rects2[0]), ("Men", "Women"))

rects2 = ax2.bar(ind + width, womenMeans, width, color="y", yerr=womenStd)


def autolabel(rects):
    # attach some text labels
    for rect in rects:
        height = rect.get_height()
        ax.text(
            rect.get_x() + rect.get_width() / 2.0,
            1.05 * height,
            "%d" % int(height),
            ha="center",
            va="bottom",
        )


fig.canvas.mpl_connect("button_press_event", press)
fig.canvas.mpl_connect("button_release_event", release)
fig.canvas.mpl_connect("key_press_event", keypress)
fig.canvas.mpl_connect("key_release_event", keyup)
fig.canvas.mpl_connect("motion_notify_event", motionnotify)
fig.canvas.mpl_connect("resize_event", resize)
fig.canvas.mpl_connect("scroll_event", scroll)
fig.canvas.mpl_connect("figure_enter_event", figure_enter)
fig.canvas.mpl_connect("figure_leave_event", figure_leave)
fig.canvas.mpl_connect("close_event", close)

fig2.canvas.mpl_connect("button_press_event", press)
fig2.canvas.mpl_connect("button_release_event", release)
fig2.canvas.mpl_connect("key_press_event", keypress)
fig2.canvas.mpl_connect("key_release_event", keyup)
fig2.canvas.mpl_connect("motion_notify_event", motionnotify)
fig2.canvas.mpl_connect("resize_event", resize)
fig2.canvas.mpl_connect("scroll_event", scroll)
fig2.canvas.mpl_connect("figure_enter_event", figure_enter)
fig2.canvas.mpl_connect("figure_leave_event", figure_leave)
fig2.canvas.mpl_connect("close_event", close)

canvas = fig.canvas
# canvas.blit(Bbox(np.array([[0, 0], [400, 400]], np.int32)))
canvas2 = fig2.canvas


def callback(instance):
    autolabel(rects1)
    autolabel(rects2)
    canvas.draw()


class MatplotlibTest(App):
    title = "Matplotlib Test"

    def build(self):
        fl = BoxLayout(orientation="vertical")
        a = Button(text="press me", height=40, size_hint_y=None)
        a.bind(on_press=callback)
        nav1 = NavigationToolbar2Kivy(canvas)
        nav2 = NavigationToolbar2Kivy(canvas2)
        fl.add_widget(nav1.actionbar)
        fl.add_widget(canvas)
        fl.add_widget(nav2.actionbar)
        fl.add_widget(canvas2)
        fl.add_widget(a)
        return fl


if __name__ == "__main__":
    MatplotlibTest().run()
